internal/procfs/procfs_linux.go

```diff
@@ -0,0 +1,88 @@
+package procfs
+
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+// PidOf finds process(es) with a specified name (regexp match)
+// and return their pid(s)
+func PidOf(name string) ([]int, error) {
+	if len(name) == 0 {
+		return []int{}, fmt.Errorf("name should not be empty")
+	}
+	re, err := regexp.Compile("(^|/)" + name + "$")
+	if err != nil {
+		return []int{}, err
+	}
+	return getPids(re), nil
+}
+
+func getPids(re *regexp.Regexp) []int {
+	pids := []int{}
+	filepath.Walk("/proc", func(path string, info os.FileInfo, err error) error {
```

Comment:
more worried if the `pidof` is available on every platform and if it is reachable from the daemon (like is in the $PATH) etc

Reply:
I'm not sure `Walk` is really what we want here since it traverses every dir recursively.

What do you think about some either hitting the systemd-resolved API or checking `/etc/resolv.conf` for `127.0.0.53` instead?

Reply:
> more worried if the pidof is available on every platform and if it is reachable from the daemon (like is in the $PATH) etc

Good point; looking at http://man7.org/linux/man-pages/man1/pidof.1.html, it's part of https://gitlab.com/procps-ng/procps. I _think_ it's fairly common to be installed (at least on common distros), but we may want to check.

I had the same concerns as @cpuguy83; recursively traversing `/proc` just for detection looks like a lot of overhead. Not sure if string-matching for `127.0.0.53` in `/etc/resolv.conf` is a good solution though; sounds a bit brittle.

We should consider to what extend we want the automatic-detection to be "perfect", and possibly add a manual configuration, as discussed in  https://github.com/docker/libnetwork/pull/2232#issuecomment-404924431

> (...) is checking for a particular binary running the right way to do this? K8s is mainly doing it to issue a warning so that's "safe" in a sense. But transparently substituting in libnetwork doesn't seem robust. Maybe we should do what k8s is doing even more so: require the user to provide an option for a default resolv.conf file when starting up the engine. (...)

Reply:
I opened a PR on k8s repo here to optimize the pid lookup: https://github.com/kubernetes/kubernetes/pull/66367

Reply:
@thaJeztah being in the config, does not automatically exposing it as an option? so if the user is not satisfied on the auto selection can specify it?

Reply:
@cpuguy83 will wait for that PR to be merged and will update, thanks for opening it